src/exodus/sigmadb.cpp

```diff
@@ -252,85 +352,107 @@ void CMPMintList::DeleteAll(int startBlock)
     auto sequenceKey = CreateSequenceKey(lastSequence);
 
     auto it = NewIterator();
     it->Seek(GetSlice(sequenceKey));
 
-    // Get the list of mints to delete.
-    struct DeleteEntry {
-        std::string sequenceKey;
-        std::string mintKey;
-        PropertyId property;
-        DenominationId denomination;
-        MintGroupId group;
-        MintGroupIndex index;
-        SigmaPublicKey pubKey;
-        int block;
-    };
-
-    std::vector<DeleteEntry> deletes;
-
+    leveldb::WriteBatch batch;
+    std::vector<function<void()>> defers; // functions to be called after delete whole keys
     for (; it->Valid() && IsSequenceEntry(it.get()); it->Prev()) {
 
-        // Load mint data.
-        std::string mintData;
+        CDataStream deserialized(
+            it->value().data(),
+            it->value().data() + it->value().size(),
+            SER_DISK, CLIENT_VERSION
+        );
 
-        auto status = pdb->Get(readoptions, it->value(), &mintData);
-        if (!status.ok()) {
-            throw std::runtime_error("DeleteAll() : fail to read mint from sequence");
-        }
-
-        // Check if it need to delete then add to the list.
-        DeleteEntry entry;
-
-        std::tie(entry.pubKey, entry.block) = ParseMint(mintData);
+        // Check if it need to delete then push it to batch
+        History entry;
+        deserialized >> entry;
 
         if (entry.block < startBlock) {
             // We iterate in the latest to oldest that mean we can stop as soon as we found it block number is lower
             // than theshold.
             break;
         }
 
-        if (!ParseMintKey(it->value(), entry.property, entry.denomination, entry.group, entry.index)) {
-            throw std::runtime_error("Found invalid mint key");
+        // intentionally using if instead of switch to separate scope
+        if (entry.op == OpCode::StoreMintKey) {
+            auto key = GetSlice(entry.data);
+
+            // retrieve meta data of mint
+            uint32_t propertyId;
+            uint8_t denomination;
+            uint32_t groupId;
+            uint16_t count;
+            if (!ParseMintKey(key, propertyId, denomination, groupId, count)) {
+                throw std::runtime_error("fail to parse mint key");
+            }
+
+            // get commitment
+            std::string data;
+            auto status = pdb->Get(readoptions, key, &data);
+            if (!status.ok()) {
+                throw std::runtime_error("fail to get mint");
+            }
+            CDataStream pubkeyDeserialized(data.data(), data.data() + data.size(), SER_DISK, CLIENT_VERSION);
+            SigmaPublicKey pub;
+            pubkeyDeserialized >> pub;
+
+            // function to trigger event
+            defers.push_back([this, propertyId, denomination, pub]() {
+                MintRemoved(propertyId, denomination, pub);
+            });
+
+            batch.Delete(GetSlice(entry.data));
```

Comment:
I'm pretty sure this is a bug due to `batch` should not copy the actual key that they need to delete.

Reply:
after checking the implementation, the content is copied.